command/agent/http.go

```diff
@@ -75,89 +76,111 @@ type HTTPServer struct {
 	wsUpgrader *websocket.Upgrader
 }
 
-// NewHTTPServer starts new HTTP server over the agent
-func NewHTTPServer(agent *Agent, config *Config) (*HTTPServer, error) {
-	// Start the listener
-	lnAddr, err := net.ResolveTCPAddr("tcp", config.normalizedAddrs.HTTP)
-	if err != nil {
-		return nil, err
-	}
-	ln, err := config.Listener("tcp", lnAddr.IP.String(), lnAddr.Port)
-	if err != nil {
-		return nil, fmt.Errorf("failed to start HTTP listener: %v", err)
-	}
+// NewHTTPServers starts an HTTP server for every address.http configured in
+// the agent.
+func NewHTTPServers(agent *Agent, config *Config) ([]HTTPServer, error) {
+	var srvs []HTTPServer
+	var error error
 
-	// If TLS is enabled, wrap the listener with a TLS listener
-	if config.TLSConfig.EnableHTTP {
-		tlsConf, err := tlsutil.NewTLSConfiguration(config.TLSConfig, config.TLSConfig.VerifyHTTPSClient, true)
+	// Start the listener
+	for _, addr := range config.normalizedAddrs.HTTP {
+		lnAddr, err := net.ResolveTCPAddr("tcp", addr)
 		if err != nil {
-			return nil, err
+			multierror.Append(error, err)
+			continue
 		}
-
-		tlsConfig, err := tlsConf.IncomingTLSConfig()
+		ln, err := config.Listener("tcp", lnAddr.IP.String(), lnAddr.Port)
 		if err != nil {
-			return nil, err
+			error = multierror.Append(error, fmt.Errorf("failed to start HTTP listener: %v", err))
+			continue
 		}
-		ln = tls.NewListener(tcpKeepAliveListener{ln.(*net.TCPListener)}, tlsConfig)
-	}
 
-	// Create the mux
-	mux := http.NewServeMux()
+		// If TLS is enabled, wrap the listener with a TLS listener
+		if config.TLSConfig.EnableHTTP {
+			tlsConf, err := tlsutil.NewTLSConfiguration(config.TLSConfig, config.TLSConfig.VerifyHTTPSClient, true)
+			if err != nil {
+				error = multierror.Append(error, err)
+				continue
+			}
 
-	wsUpgrader := &websocket.Upgrader{
-		ReadBufferSize:  2048,
-		WriteBufferSize: 2048,
-	}
+			tlsConfig, err := tlsConf.IncomingTLSConfig()
+			if err != nil {
+				error = multierror.Append(error, err)
+				continue
+			}
```

Comment:
Can we reuse the TLS configuration across listeners?

Reply:
Let me test this locally with a self signed cert and let you know

Reply:
:+1: setup using https://learn.hashicorp.com/tutorials/nomad/security-enable-tls and ran `./bin/nomad agent -dev -config config.hcl` with the following config:
```hcl
addresses {
  http = "127.0.0.1 127.0.0.2"
}


# Require TLS
tls {
  http = true
  rpc  = true

  ca_file   = "nomad-ca.pem"
  cert_file = "server.pem"
  key_file  = "server-key.pem"
}
```

and navigated to .1 and .2 with no problems.